beacon_node/store/src/config.rs

```diff
@@ -210,28 +225,70 @@ impl StoreItem for OnDiskStoreConfig {
     }
 
     fn as_store_bytes(&self) -> Vec<u8> {
-        self.as_ssz_bytes()
+        match self {
+            // V0 is not prefixed explicitly, but unless the user chose a crazy configuration the
+            // serialized value will have a leading 0 byte.
+            OnDiskStoreConfig::V0(value) => value.as_ssz_bytes(),
+            OnDiskStoreConfig::V1(value) => {
+                let mut out = value.as_ssz_bytes();
+                out.insert(0, 1);
+                out
+            }
+        }
     }
 
     fn from_store_bytes(bytes: &[u8]) -> Result<Self, Error> {
-        Ok(Self::from_ssz_bytes(bytes)?)
+        let first_byte =
+            *bytes
+                .first()
+                .ok_or(Error::ConfigError(StoreConfigError::InvalidVersionByte(
+                    None,
+                )))?;
+
+        match first_byte {
+            // Decode all payload, as there's no prefix byte
+            // NOTE: V1 config can never be deserialized as a V0 because the minimum length of its
+            // serialization is: 1 prefix byte + 1 offset (OnDiskStoreConfigV1 container) +
+            // 1 offset (HierarchyConfig container) = 9.
+            0 => Ok(Self::V0(OnDiskStoreConfigV0::from_ssz_bytes(bytes)?)),
+            // Skip first prefix byte
+            1 => Ok(Self::V1(OnDiskStoreConfigV1::from_ssz_bytes(&bytes[1..])?)),
+            other => Err(Error::ConfigError(StoreConfigError::InvalidVersionByte(
+                Some(other),
+            ))),
+        }
     }
 }
 
 #[cfg(test)]
 mod test {
     use super::*;
     use crate::{metadata::STATE_UPPER_LIMIT_NO_RETAIN, AnchorInfo, Split};
+    use ssz::DecodeError;
     use types::{Hash256, Slot};
 
     #[test]
     fn check_compatibility_ok() {
         let store_config = StoreConfig {
             ..Default::default()
         };
-        let on_disk_config = OnDiskStoreConfig {
+        let on_disk_config = OnDiskStoreConfig::V1(OnDiskStoreConfigV1 {
             hierarchy_config: store_config.hierarchy_config.clone(),
+        });
+        let split = Split::default();
+        assert!(store_config
+            .check_compatibility(&on_disk_config, &split, None)
+            .is_ok());
+    }
+
+    #[test]
+    fn check_compatibility_after_migration() {
+        let store_config = StoreConfig {
+            ..Default::default()
         };
+        let on_disk_config = OnDiskStoreConfig::V0(OnDiskStoreConfigV0 {
+            slots_per_restore_point: 0,
```

Comment:
```suggestion
            slots_per_restore_point: 8192,
```

I think we should use a valid config here, `slots_per_restore_point == 0` was never allowed.

Reply:
Fixed in e5661720685236e73c6fee55998f15ba44a9ed88